### ISSUE TITLE
There is no method to create a train / test pair

### ISSUE THREAD

**Issue by mbaudin47** [openturns/openturns#1425]
In order to create a metamodel, we often use cross-validation where one part of the sample is used for training and the other part for testing. The `split` method of the `Sample` does not allow to shuffle the data. Since two samples are involved, this should be a static method of the `Sample` class. 

Below is an implementation. 

```
def SampleTrainTestSplit(xSample, ySample, 
                        testSize = None, 
                        shuffle = True):
    """
    Split the sample into train and test samples. 

    Parameters
    ----------
    xSample : ot.Sample
        The input sample.

    ySample : ot.Sample
        The output sample.

    testSize : int
        The size of the test sample. 
        If None, set to 0.25 * sampleSize. 
        
    shuffle : bool
        If True, then shuffles the sample. 
    
    Returns
    -------
    xSampleTrain : ot.Sample
        The training input sample.  

    xSampleTest : ot.Sample
        The test input sample.  

    ySampleTrain : ot.Sample
        The training output sample.  

    ySampleTest : ot.Sample
        The test output sample.  
    """
    # Compute train and test sizes
    sampleSize = xSample.getSize()
    if testSize is None:
        testSize = int(0.25 * sampleSize)
    trainSize = sampleSize - testSize
    # Random permutation
    if shuffle:
        distribution = ot.KPermutationsDistribution(sampleSize, sampleSize)
        floatIndices = distribution.getSample(1)
        indices = [int(floatIndices[0,i]) for i in range(sampleSize)]
        xArray = np.array(xSample)
        x = ot.Sample(xArray[indices])
        yArray = np.array(ySample)
        y = ot.Sample(yArray[indices])
    else:
        x = ot.Sample(xSample)
        y = ot.Sample(ySample)
    # Split
    xSampleTrain = x[:trainSize, :]
    ySampleTrain = y[:trainSize, :]
    xSampleTest = x[trainSize:, :]
    ySampleTest = y[trainSize:, :]
    return xSampleTrain, xSampleTest, ySampleTrain, ySampleTest
```

Here are two tests.

```
    def test_SampleTrainTestSplitNoShuffle(self):
        sampleSize = 12
        inputDistribution = ot.ComposedDistribution([ot.Uniform()]*3)
        outputDistribution = ot.ComposedDistribution([ot.Uniform()]*2)
        xSample = inputDistribution.getSample(sampleSize)
        ySample = outputDistribution.getSample(sampleSize)
        xSampleTrain, xSampleTest, \
            ySampleTrain, ySampleTest \
                = otfrasil.SampleTrainTestSplit(xSample, ySample, 
                                                shuffle = False)
        assert(xSampleTrain.getSize() == 9)
        assert(xSampleTest.getSize() == 3)
        assert(xSampleTrain.getSize() == 9)
        assert(ySampleTest.getSize() == 3)
        npt.assert_allclose(xSampleTrain, xSample[:9])
        npt.assert_allclose(xSampleTest, xSample[9:])
        npt.assert_allclose(ySampleTrain, ySample[:9])
        npt.assert_allclose(ySampleTest, ySample[9:])

    def test_SampleTrainTestSplitShuffle(self):
        sampleSize = 12
        inputDistribution = ot.ComposedDistribution([ot.Uniform()]*3)
        outputDistribution = ot.ComposedDistribution([ot.Uniform()]*2)
        xSample = inputDistribution.getSample(sampleSize)
        ySample = outputDistribution.getSample(sampleSize)
        xSampleTrain, xSampleTest, \
            ySampleTrain, ySampleTest \
                = otfrasil.SampleTrainTestSplit(xSample, ySample)
        assert(xSampleTrain.getSize() == 9)
        assert(xSampleTest.getSize() == 3)
        assert(xSampleTrain.getSize() == 9)
        assert(ySampleTest.getSize() == 3)
        # Check that sort samples match
        xSample = xSample.sort()
        fullXSample = ot.Sample(xSampleTrain)
        for p in xSampleTest:
            fullXSample.add(p)
        fullXSample = fullXSample.sort()
        npt.assert_allclose(xSample, fullXSample)
        ySample = ySample.sort()
        fullYSample = ot.Sample(ySampleTrain)
        for p in ySampleTest:
            fullYSample.add(p)
        fullYSample = fullYSample.sort()
        npt.assert_allclose(ySample, fullYSample)
```

**Comment by mbaudin47** [comment]
I am creating a draft of a class and will integrate it when possible.

**Comment by mbaudin47** [comment]
@regislebrun : I am working on the right location and design of this service.

1) This class cannot be a static method of the `Sample` class, because `Sample` is in Base/Stat. However, the method involves a shuffling algorithm which requires `KPermutationsDistribution` from Uncertainty/Distribution. 

2) This cannot be a static method, because it returns 4 variables (xSampleTrain, xSampleTest, ySampleTrain, ySampleTest), which can be done in Python but not in C++. 

3) Could it be implemented as an Algorithm? Then the use case would be:
```
algo = TrainTestSplitAlgorithm(xSample, ySample, shuffle = False)
algo.run()
result = algo.getResult()
xSampleTrain = result.getXSampleTrain()
xSampleTest = result.getXSampleTest()
ySampleTrain = result.getYSampleTrain()
ySampleTest = result.getYSampleTest()
```
The TrainTestSplitResult class would have ctor:
```
result = TrainTestSplitResult(xSampleTrain, xSampleTest, ySampleTrain, ySampleTest)
```
This sounds like a feasible way to me. Is that correct?

**Comment by sofianehaddad** [comment]
@mbaudin47 I fully agree that it is important to have such functionality. IMHO the work could not be achieved as proposed because we have some incoherence in the API

We have some classes for model selection
(http://openturns.github.io/openturns/latest/user_manual/response_surface/_generated/openturns.CorrectedLeaveOneOut.html#openturns.CorrectedLeaveOneOut, http://openturns.github.io/openturns/latest/user_manual/response_surface/_generated/openturns.KFold.html?highlight=kfold). These classes `KFold` and `CorrectedLeaveOneOut` are *chaos oriented* and **should** be more generic. Until now, we didn't get in mind having features for model selection/cross validation. If so, designing `train_test_split` as proposed will enrich the library with a new feature but it is a wrong way as the new class has no connection with others.

Thus the work is not so easy and we should redesign the API in order to have more generic `model_selection` features and derives the specificities for Chaos usage. What do you think ?

As an example of a global submodule, have a look at https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.KFold.html and 
https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.ShuffleSplit.html#sklearn.model_selection.ShuffleSplit
